PEP 722: Dependency specification for single-file scripts (#3210)
python/peps: pep-0722.rst

single-file Python scripts (which are often kept in some sort of shared
directory). This PEP offers a solution for that use case, by storing the
dependencies in the script itself.

This proposal is (a subset of) behaviour that already exists in the ``pipx`` and
``pip-run`` tools, so it is simply formalising existing behaviour, rather than
defining a brand new capability.

Rationale
=========

In general, the Python packaging ecosystem is focused on "projects", which are
structured around a directory containing your code, some metadata and any
supporting files such as tests, automation, etc. However, beginners (and quite a
few more experienced developers) will often *start* a project by just opening up
an editor and writing a simple script. In some cases, the project might outgrow
the "simple script" phase and get restructured into a more conventional project
directory, but not all do, and even those that do still need to be usable in the
"simple script" phase.

These days, the idea that "simple scripts can just use the stdlib" is becoming
less and less practical, as the quantity, quality and usability of libraries on

> PyPI steadily increases. For example, using the stdandard ``urllib`` library

Reviewer: ```suggestion
PyPI steadily increases. For example, using the standard ``urllib`` library
```

rather than something like ``requests`` or ``httpx`` makes correct code
significantly harder to write.

Having to consider "uses 3rd party libraries" as the break point for moving to a
"full scale project" is impractical, so this PEP is designed to allow a project
to use external libraries while still remaining as a simple, standalone script.

Of course, *declaring* your dependencies isn't sufficient by itself. You need to
install them (probably in some sort of virtual environment) so that they are
available when the script runs. This PEP does not cover environment management,
as tools like `pip-run <https://pypi.org/project/pip-run/>`__ and `pipx
<https://pypi.org/project/pipx/>`__ already offer that ability. But by
standardising the means of declaring dependencies, this PEP allows scripts to
remain tool-independent.

Specification
=============

Any Python script may contain a *dependency block*, which is a specially
structured comment block. The format of the dependency block is as follows:

> * A single comment line containing the (case sensitive) text "Requirements:"

Reviewer 2: Case-sensitivity is important, I wouldn't put it in a paranthetical aside:

```suggestion
* A single comment line containing the case sensitive text "Requirements:"
```

* A series of comment lines containing :pep:`508` requirement specifiers.
* An empty comment or blank line.

To be recognised as a "comment line", the line must start with a ``#`` symbol.
Leading and trailing whitespace on the comment line is ignored.

The dependency block may occur anywhere in the file. There MUST only be a single
dependency block in the file - tools consuming dependency data MAY simply
process the first dependency block found. This avoids the need for tools to
process more data than is necessary. Stricter tools MAY, however, fail with an
error if multiple dependency blocks are present.

Example
-------

The following is an example of a script with an embedded dependency block::

    #!/usr/bin/env python

    # In order to run, this script needs the following 3rd party libraries
    #
    # Requirements:

>     #    requests

Reviewer 3: Perhaps these requirements should follow a similar declarative syntax to pyproject.toml, specifically to allow for versioning? Accepting a dependency, without hinting as to the potential required version makes ongoing maintenance more difficult, IMHO.

Author: Discussion should happen on the Python Discourse thread noted in the PEP header. Reviews here should *only* be for typos or procedural matters. Please raise your point on Discourse.

Unless you just mean you'd like me to include a version spec in the example, something like `requests>=2.30`? I assumed that the statement that PEP 508 requirements were allowed made that clear, and I prefer having realistic examples. But if there's wider support for doing that, I can - I don't intend to cover all of PEP 508, though...

    #    rich

    import requests
    from rich.pretty import pprint

    resp = requests.get("https://peps.python.org/api/peps.json")
    data = resp.json()
    pprint([(k, v["title"]) for k, v in data.items()][:10])

Backwards Compatibility
=======================

As the dependency data is recorded in the form of a structured comment, this is

> compatible with any existing code.

Reviewer 2: Perhaps stretching the bounds of plausibility, but if someone has a file with exactly a structure as defined in this PEP, tools could unexpectedly start installing packages when the script runs. A potential compatability concern, though I don't think particularly major.

This would typically involve storing the dependencies as a (runtime) list
variable with a conventional name, such as::

    __requires__ = [
        "requests",
        "click",
    ]

This has a number of problems compared to the proposed solution.

1. The consumer has to parse arbitrary Python code, which almost certainly means
   using the stdlib AST module, making it much harder for non-Python code to
   read the data, as well as making Python code that does so significantly more
   complex.
2. Python syntax changes every version. While the requirement data only uses a
   simple subset, the full file still needs to be parsed to *find* the
   requirement data.
3. This would reserve a specific global name (``__requires__``) in the above,
   potentially clashing with user code.
4. Users could assume that the value can be manipulated at runtime, and would
   get unexpected results if they tried to do so.

Furthermore, there is no known use case where being able to read the list of
requirements at runtime is needed.

It is worth noting, though, that the ``pip-run`` utility does implement (an
extended form of) this approach. See `here <pip-run issue_>`_ for further

> discussion.

Reviewer 4: Minor: best to avoid using just `here` for link texts.

Reviewer 2: A suggested rephrase from Éric's point:

```suggestion
extended form of) this approach.
(`Further discussion <pip-run issue_>`_) is available on
their design of 'embedded requirements'.
```

The pip requirements file format allows a lot more flexibility than a simple
list of requirements - it allows pip options, including specification of
non-standard indexes. The requirements format is not standardised, though, and
never will be in its current form, as it includes a lot of pip-specific
functionality.

This proposal deliberately does not try to replicate the full feature set of a
requirements file. It would be possible to implement "some" features, for
example being able to add extra index locations. However, it is difficult to
know where to draw the line, and not all consumers of this data may be passing
the dependencies to pip (for example, a script vulnerability scanner).

If a script needs the full requirements file capabilities, it can be shipped
with an accompanying requirements file. While this means the code can no longer
be shipped as a single file, it has probably reached a point of complexity where
"having everything in a single file" is no longer an appropriate goal anyway.

There is more discussion of this point in `the previously mentioned pip-run
issue <pip-run issue_>`_.

What about local dependencies?
------------------------------

> :pep:`508` does not allow local directories or files as dependecy specifiers.

Reviewer: ```suggestion
:pep:`508` does not allow local directories or files as dependency specifiers.
```

PEP: 722
Title: Dependency specification for single-file scripts
Author: Paul Moore <p.f.moore@gmail.com>
PEP-Delegate: Brett Cannon <brett@python.org>
Discussions-To: https://discuss.python.org/t/29905
Status: Draft
Type: Standards Track
Topic: Packaging
Content-Type: text/x-rst
Created: 19-Jul-2023
Post-History: `19-Jul-2023 <https://discuss.python.org/t/29905>`__

Abstract
========

This PEP specifies a format for including 3rd-party dependencies in a
single-file Python script.

Motivation
==========

Not all Python code is structured as a "project", in the sense of having its own
directory complete with a ``pyproject.toml`` file, and being built into an
installable distribution package. Python is also routinely used as a scripting
language, with Python scripts as a (better) alternative to shell scripts, batch

> files, etc. When used to create scipts, Python code is typically stored as a

Author: Thanks. Is there a way to do that on Windows? I can sort of work out from looking at the makefile what command I'd use, but I didn't even know what make subcommands existed 🙁

file. This is deliberate, as Python syntax changes over time, so attempting to
parse the script as Python code would require choosing a specific version of
Python syntax. Also, it is likely that at least some tools will not be written
in Python, and expecting them to implement a Python parser is too much of a
burden.

However, to avoid needing changes to core Python, the format is designed to
appear as comments to the Python parser. It is possible to write code where a
metadata block is *not* interpreted as a comment (for example, by embedding it
in a Python multi-line string), but such uses are discouraged and can easily be
avoided assuming you are not deliberately trying to create a pathological
example.

A `review <language survey_>`_ of how other languages allow scripts to specify
their dependencies shows that a "structured comment" like this is a
commonly-used approach.

Specification
=============

Any Python script may contain one or more *metadata blocks*. Metadata blocks are
identified by reading the script *as a text file* (i.e., the file is not parsed
as Python source code), looking for contiguous blocks of lines that start with

> the identifying chacaters ``##``. Whitespace is not allowed before the

Reviewer 5: Typo here, `chacaters`.

Author: Sigh. Thanks. Someday I'll see if I work out how to implement a Windows-compatible version of the `make spellcheck` command.

Reviewer: `make spellcheck` didn't pick this one up, but to mirror the commands from the `Makefile`, it's essentially:

```sh
# Install
python3 -m pip install pre-commit

# Run
pre-commit run --all-files --hook-stage manual codespell
```

Or:
```sh
# Install
python3 -m pip install codespell

# Run, but it might pick up junk from .tox, .venv, other non-Git files
codespell

# Run on a file
codespell pep-0722.rst
```

identifying ``##``. More than one metadata block may exist in a Python file.

Tools reading dependency blocks MAY respect the standard Python encoding
declaration. If they choose not to do so, they MUST process the file as UTF-8.

> Within a metadata block, whitespace at the start and end of each line is

Reviewer 2: Does this mean whitespace after the ``##`` and before any other text?

Author: Yes. It’s explicit in the reference implementation. I was trying not to be too wordy here. I’ll add a note to make it clear.

the current status quo, though, where running a script without its dependencies
present will give an error.

In general, it is assumed that if a beginner is given a script with dependencies
(regardless of whether they are specified in a dependency block), the person
supplying the script should explain how to run that script, and if that involves
using a script runner tool, that should be noted.

Recommendations
===============

This section is non-normative and simply describes "good practices" when using
metadata blocks.

Scripts should, in general, place metadata blocks at the top of the file,
either immediately after any shebang line, or straight after the script
docstring. In particular, the metadata block should always be placed before
any executable code in the file. This makes it easy for the human reader to
locate the metadata block, and allows tools to only read the minimum necessary
to identify them.

Reference Implementation
========================

Code to implement this proposal in Python is fairly straightforward, so the
reference implementation can be included here.

A parser that reads *only* the script dependency metadata.

> .. code-block: python

Reviewer 2: ```suggestion
.. code:: python
```

   import tokenize
   from packaging.requirements import Requirement

   DEPENDENCY_BLOCK_MARKER = "Script Dependencies:"

   def read_dependency_block(filename):
      # Use the tokenize module to handle any encoding declaration.
      with tokenize.open(filename) as f:
         for line in f:
               if line.startswith("##"):
                  line = line[2:].strip()
                  if line == DEPENDENCY_BLOCK_MARKER:
                     for line in f:
                        if not line.startswith("##"):
                              break
                        line = line[2:].strip()
                        if not line:
                              continue
                        # Try to convert to a requirement. This will raise
                        # an error if the line is not a PEP 508 requirement
                        yield Requirement(line)
                     break

A full metadata block parser that returns all metadata blocks in a script.

> .. code-block: python

Reviewer 2: ```suggestion
.. code:: python
```

First of all, the only practical choice for an alternative format is TOML.
Python packaging has standardised on TOML for structured data, and using a
different format, such as YAML or JSON, would add complexity and confusion for
no real benefit.

So the question is essentially, "why not use TOML?"

The key idea behind the "metadata block" format is to define something that
reads naturally as a comment in the script. Dependency data is useful both for
tools and for the human reader, so having a human readable format is beneficial.
On the other hand, TOML of necessity has a syntax of its own, which distracts
from the underlying data.

It is important to remember that developers who *write* scripts in Python are
often *not* experienced in Python, or Python packaging. They are often systems
administrators, or data analysts, who may simply be using Python as a "better
batch file". For such users, the TOML format is extremely likely to be
unfamiliar, and the syntax will be obscure to them, and not particularly
intuitive. Such developers may well be copying dependency specifiers from
sources such as Stack Overflow, without really understanding them. Having to

> embed such a requirement into a TOML structure is an additional complexity - and

Reviewer 2: Minor typographical suggestion, for a 'nicer' hypen.
```suggestion
embed such a requirement into a TOML structure is an additional complexity -- and
```

it is important to remember that the goal here is to make using 3rd party
libraries *easy* for such users.

Furthermore, TOML, by its nature, is a flexible format intended to support very
general data structures. There are *many* ways of writing a simple list of
strings in it, and it will not be clear to inexperienced users which form to use.

And finally, there will be tools that expect to *write* dependency data into

> scripts - for example, an IDE with a feature that automatically adds an import

Reviewer 2: Minor typographical suggestion, for a 'nicer' hypen.
```suggestion
scripts -- for example, an IDE with a feature that automatically adds an import
```

scripts, and metadata for distribution packages (as defined in
``pyproject.toml``). In both cases, dependencies are given in "abstract" form,
without specifying how they are obtained.

Some tools that use the dependency information may, of course, need to locate
concrete dependency artifacts - for example if they expect to create an
environment containing those dependencies. But the way they choose to do that
will be closely linked to the tool's UI in general, and this PEP does not try to
dictate the UI for tools.

There is more discussion of this point, and in particular of the UI choices made
by the ``pip-run`` tool, in `the previously mentioned pip-run issue <pip-run
issue_>`_.

What about local dependencies?
------------------------------

These can be handled without needing special metadata and tooling, simply by
adding the location of the dependencies to ``sys.path``. This PEP simply isn't
needed for this case. If, on the other hand, the "local dependencies" are actual
distributions which are published locally, they can be specified as usual with a
:pep:`508` requirement, and the local package index specified when running a
tool by using the tool's UI for that.

> Open Issues
> ===========
> 
> None at this point.
> 
> 

Reviewer 2: Perhaps remove the section if there are no open issues?

```suggestion
```

Reviewer 4: The presense of an empty list is saying something that its absence wouldn’t! 🙂

Author: I agree with @merwok on this one.

> Footnotes

Reviewer 2: We automatically remove 'references' sections with no content bar targets.

```suggestion
References
```

Author: I don’t understand this comment sorry. Footnotes was the header in the template. But I’m fine with changing it if you think it helps.

Author: TBH, it's not that important. I certainly don't have any problem with accepting this suggestion, and I'd rather get this merged sooner rather than later, so that the PEP is available "for real", rather than just as a PR. So I'll commit this suggestion and then merge once CI passes. Hope that's OK.

Reviewer 4: My understanding: if this section is named References and contains only reST link targets, then the data will be used to create the link elements and the section will be removed from the rendered contents, instead of being there and looking strange (since PEPs now use real links and not paper-style `[1]` references that lead to end of document)

Author: Yeah, just spotted that, thanks 🙂 And thanks for the explanation.

Reviewer 2: Sorry for being a bit too brief -- @merwork is exactly right!

A

    #
    ## Script Dependencies:
    ##    requests
    ##    rich

    import requests
    from rich.pretty import pprint

    resp = requests.get("https://peps.python.org/api/peps.json")
    data = resp.json()
    pprint([(k, v["title"]) for k, v in data.items()][:10])

Backwards Compatibility
=======================

As metadata blocks take the form of a structured comment, they can be added
without altering the meaning of existing code.

It is possible that a comment may already exist which matches the form of a
metadata block. While the use of a double ``#`` prefix is intended to minimise
this risk, it is still possible.

Because tools must ignore unrecognised metadata types, the only potential issue
we need to consider is script dependencies. In that case, a tool might read the
wrong dependencies. In practice, though, this is unlikely to happen, as (a) the
header text (``Script Dependencies:``) is fairly unusual, and (b) any following
lines are unlikely to conform to :pep:`508` unless they *are* dependencies.

In the rare case where an existing comment would be interpreted incorrectly as a
dependency block, this can be addressed by adding an actual dependency block

> (which can be empty if the script has no dependencies) earlier in the code.

Reviewer 6: This would run afoul of the earlier rule, “There MUST only be a single ``Script Dependencies`` block in the file.”

Author: Good point. I'll relax the MUST, as all it's actually intended to say is that tools are allowed to ignore anything except the first one.
